src/contracts/Executor.sol

```diff
@@ -18,95 +22,135 @@ contract Executor is Context {
     using ECDSA for bytes32;
 
     IBridge public immutable _bridge;
+    uint8 public immutable _domainID;
+    uint256 public _slotIndex;
+    IAccessControlSegregator public _accessControl;
 
+    // securityModel => block header storage address
+    mapping(uint8 => address) public _securityModels;
     // origin domainID => nonces set => used deposit nonces
     mapping(uint8 => mapping(uint256 => uint256)) public usedNonces;
+    //  origin domainID => router address
+    mapping(uint8 => address) public _originDomainIDToRouter;
 
     struct Proposal {
         uint8 originDomainID;
+        uint8 securityModel;
         uint64 depositNonce;
         bytes32 resourceID;
         bytes data;
+        bytes[] storageProof;
     }
 
-    event ProposalExecution(uint8 originDomainID, uint64 depositNonce, bytes32 dataHash, bytes handlerResponse);
-    event FailedHandlerExecution(bytes lowLevelData, uint8 originDomainID, uint64 depositNonce);
+    event ProposalExecution(uint8 originDomainID, uint64 depositNonce);
+    event FeeRouterChanged(uint8 originDomainID, address newRouter);
+    event SlotIndexChanged(uint256 slotIndex);
 
     error EmptyProposalsArray();
     error BridgeIsPaused();
+    error AccessNotAllowed(address sender, bytes4 funcSig);
+    error TransferHashDoesNotMatchSlotValue(bytes32 transferHash);
+
+    modifier onlyAllowed() {
+        _onlyAllowed(msg.sig, _msgSender());
+        _;
+    }
 
     modifier whenBridgeNotPaused() {
         if (_bridge.paused()) revert BridgeIsPaused();
         _;
     }
 
-    constructor(address bridge) {
+    function _onlyAllowed(bytes4 sig, address sender) private view {
+        if (!_accessControl.hasAccess(sig, sender)) revert AccessNotAllowed(sender, sig);
+    }
+
+    constructor(
+        uint8 originDomainID,
```

Comment:
This should either be an array or you should remove it and just use the function.